### PR TITLE
Add support for STM32L432KC and Nucleo L432KC board

### PR DESCRIPTION
**Pull request by superna9999**
The STM32L4x2 SoC serie is a lowered down version of the STM32L4x6 SoCs.

This PR adds support for the STM32L432KC SoC and support for the Nucleo L432KC from the Nucleo-32 boards serie.
Documentation for the Nucleo L432KC is added based on the Nucleo L476RG format.

---
This change is [Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/19)

### CODE REVIEW

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+Hardware
+********
+
+The STM32L432KC SoC provides the following hardware IPs:
+
+- Ultra-low-power with FlexPowerControl (down to 28 nA Standby mode and 84 μA/MHz run mode)
+- Core: ARM® 32-bit Cortex®-M4 CPU with FPU, frequency up to 80 MHz, 100DMIPS/1.25DMIPS/MHz (Dhrystone 2.1)
+- Clock Sources:
+        - 32 kHz crystal oscillator for RTC (LSE)
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Lists (and sub-lists) need a blank line before and after the lists, so add a blank line before - 32 kHz

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        - Internal 16 MHz factory-trimmed RC (±1%)
+        - Internal low-power 32 kHz RC (±5%)
+        - Internal multispeed 100 kHz to 48 MHz oscillator, auto-trimmed by LSE (better than ±0.25 % accuracy)
+        - 2 PLLs for system clock, USB, audio, ADC
+- RTC with HW calendar, alarms and calibration
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: add a blank like after the (sub)list ends and before - RTC

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+- Up to 3 capacitive sensing channels: support touchkey, linear and rotary touch sensors
+- 11x timers:
+        - 1x 16-bit advanced motor-control
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: add a blank line before this sublist

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        - 1x 32-bit and 2x 16-bit general purpose
+        - 2x 16-bit basic
+        - 2x low-power 16-bit timers (available in Stop mode)
+        - 2x watchdogs
+        - SysTick timer
+- Up to 26 fast I/Os, most 5 V-tolerant
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: ... and a blank line after the sublist

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+- Memories
+        - Up to 256 KB single bank Flash, proprietary code readout protection
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Add blank line before and after the sublist... same for the sublists below

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        - 2x ultra-low-power comparators
+- 13x communication interfaces
+        - USB OTG 2.0 full-speed crystal less solution with LPM and BCD
+        - 1x SAIs (serial audio interface)
+        - 2x I2C FM+(1 Mbit/s), SMBus/PMBus
+        - 3x USARTs (ISO 7816, LIN, IrDA, modem)
+        - 2x SPIs (3x SPIs with the Quad SPI)
+        - CAN (2.0B Active)
+        - SWPMI single wire protocol master I/F
+	- IRTIM (Infrared interface)
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: indent needs to match previous line

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+The Nucleo L432KC board features an ARM Cortex-M4 based STM32L432KC MCU
+with a wide range of connectivity support and configurations. Here are
+some highlights of the Nucleo L432KC board:
+
+
+- STM32 microcontroller in UFQFPN32 package
+- Arduino Uno V3 connectivity
+- On-board ST-LINK/V2-1 debugger/programmer with SWD connector
+- Flexible board power supply:
+       - USB VBUS or external source(3.3V, 5V, 7 - 12V)
```
Comment (dbkinder) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: sublists should have a blank like before and after htem

**Review by dbkinder — CHANGES_REQUESTED** [review]
A few formatting tweaks, otherwise looks OK

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 1 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 2 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 3 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 4 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 5 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 6 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

File: boards/arm/nucleo_l432kc/doc/nucleol432kc.rst (hunk 7 above)
Comment (superna9999) [inline boards/arm/nucleo_l432kc/doc/nucleol432kc.rst]: Ok

**Review by dbkinder — COMMENTED** [review]
+1

File: dts/arm/nucleo_l432kc.fixup [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/* This file is a temporary workaround for mapping of the generated information
+ * to the current driver definitions.  This will be removed when the drivers
+ * are modified to handle the generated information, or the mapping of
+ * generated data matches the driver definitions.
+ */
+
+#define CONFIG_NUM_IRQ_PRIO_BITS	ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define CONFIG_UART_STM32_PORT_1_BASE_ADDRESS	ST_STM32_USART_40013800_BASE_ADDRESS
+#define CONFIG_UART_STM32_PORT_1_BAUD_RATE	ST_STM32_USART_40013800_BAUD_RATE
```
Comment (erwango) [inline dts/arm/nucleo_l432kc.fixup]: Rename ST_STM32_USART_40013800_BAUD_RATE as ST_STM32_USART_40013800_CURRENT_SPEED

File: dts/arm/nucleo_l432kc.fixup [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+#define CONFIG_UART_STM32_PORT_1_IRQ_PRI	ST_STM32_USART_40013800_IRQ_0_PRIORITY
+#define PORT_1_IRQ				ST_STM32_USART_40013800_IRQ_0
+
+#define CONFIG_UART_STM32_PORT_2_BASE_ADDRESS	ST_STM32_USART_40004400_BASE_ADDRESS
+#define CONFIG_UART_STM32_PORT_2_BAUD_RATE	ST_STM32_USART_40004400_BAUD_RATE
```
Comment (erwango) [inline dts/arm/nucleo_l432kc.fixup]: same here

File: dts/arm/nucleo_l432kc.fixup [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+#define CONFIG_UART_STM32_PORT_2_IRQ_PRI	ST_STM32_USART_40004400_IRQ_0_PRIORITY
+#define PORT_2_IRQ				ST_STM32_USART_40004400_IRQ_0
+
+#define CONFIG_UART_STM32_PORT_3_BASE_ADDRESS	ST_STM32_USART_40004800_BASE_ADDRESS
+#define CONFIG_UART_STM32_PORT_3_BAUD_RATE	ST_STM32_USART_40004800_BAUD_RATE
```
Comment (erwango) [inline dts/arm/nucleo_l432kc.fixup]: same here

File: dts/arm/nucleo_l432kc.fixup [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+#define CONFIG_UART_STM32_PORT_3_IRQ_PRI	ST_STM32_USART_40004800_IRQ_0_PRIORITY
+#define PORT_3_IRQ				ST_STM32_USART_40004800_IRQ_0
+
+#define CONFIG_UART_STM32_PORT_4_BASE_ADDRESS	ST_STM32_USART_40004C00_BASE_ADDRESS
+#define CONFIG_UART_STM32_PORT_4_BAUD_RATE	ST_STM32_USART_40004C00_BAUD_RATE
```
Comment (erwango) [inline dts/arm/nucleo_l432kc.fixup]: same here

**Review by erwango — APPROVED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
Sorry previous comment.  I only ment to +1 only first commit.
Removing the "approved on all changes"

File: dts/arm/nucleo_l432kc.fixup (hunk 8 above)
Comment (superna9999) [inline dts/arm/nucleo_l432kc.fixup]: Ok

**Review by galak — CHANGES_REQUESTED** [review]
The SoC and Board bits should be split, so that the board.dts, dts/arm/Makefile, fixup, are part of the patch that adds the board to Zephyr.

**Comment by superna9999** [comment]
@galak it seems the shippable script throws me errors on commits not part of this PR

**Review by galak — CHANGES_REQUESTED** [review]
So the patch sequence seems out of order here:
The board patch should be last.  I'd expect:

1. pinmux: stm32: Do not compile PORTD when not available
2. arm: stm32l4: Add configuration for STM32L432XX 
(merge dts: Add support for STM32L432)
3. pinmux: stm32: Add support for Nucleo L432KC
4. boards: arm: Add Nucleo L432KC (merge dts: Add support for Nucleo L432KC)

**Comment by superna9999** [comment]
Hmm, when pushing partial rebase, the order showed in the PR is wrong..

On the original branch (https://github.com/superna9999/zephyr/commits/stm32l432_nucleo) , the commits are:
1f9bafd boards: arm: Add support for Nucleo L432KC
b0de4a3 pinmux: stm32: Add support for Nucleo L432KC
7f2fff1 arm: stm32l4: Add configuration and dts for STM32L432XX
3106ca4 pinmux: stm32: Do not compile PORTD when not available

But the shippable status is on the good commit.

**Review by galak — APPROVED** [review]
Looks, good - let me see about dbkinder adding his approval

**Review by dbkinder — APPROVED** [review]
I guess just leaving a +1 comment doesn't work, so I'll pick the Approve button this time :+1:

**Pull request merged**